File: Data Structures & Algorithms/interleaving-string/submission-2.cpp
class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        vector<vector<bool>> dp(s1.size()+1, vector<bool>(s2.size()+1, false));
        if(s3.size() != s1.size()+s2.size()){
            return false;
        }
        dp[s1.size()][s2.size()] = true;
        for(int i = s1.size(); i>=0; i--){
            for(int j = s2.size(); j>=0; j--){
                if(i < s1.size() && dp[i+1][j] && s1[i] == s3[i+j]){
                    dp[i][j] = true;
                }
                if(j < s2.size() && dp[i][j+1] && s2[j] == s3[i+j]){
                    dp[i][j] = true;
                }
            }
        }
        return dp[0][0];
    }
};

File: Data Structures & Algorithms/generate-parentheses/submission-2.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        stack = []
        curr = []
        def dfs(openN, closeN):
            if openN == closeN == n:
                stack.append("".join(curr))
                return
            if openN < n:
                curr.append('(')
                dfs(openN+1, closeN)
                curr.pop()
            if closeN < openN:
                curr.append(')')
                dfs(openN, closeN+1)
                curr.pop()
        dfs(0, 0)
        return stack